### PR TITLE
Firebase

### PR DESCRIPTION
**Pull request by etrobert**

### CODE REVIEW

File: src/useFirestoreState.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { projectIdState, Task, TaskId } from "./atoms";
+
+type UseFirestoreState = () => {
+  addTask: (name: string) => void;
+  updateTask: (id: TaskId, task: Partial<Task>) => void;
+};
+
+const useFirestoreState: UseFirestoreState = () => {
+  const projectId = useRecoilValue(projectIdState);
+  const addTask = useCallback(
```
Comment (justinline) [inline src/useFirestoreState.ts:15]: thought: There's two implementations of `addTask` one here and one in `useGraphState` it's a bit hard to follow, is there no way to merge the hooks/state logic into one?

File: src/useFirestoreState.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    (name: string) => {
+      const ref = collection(firestore, `projects/${projectId}/tasks`);
```
Comment (justinline) [inline src/useFirestoreState.ts:17]: thought: if you're going to use firestore you should probably document the data structure so that other developers can understand it if they don't have access 🤔

File: src/useFirestoreState.ts (hunk 1 above)
Comment (etrobert) [reply src/useFirestoreState.ts:15]: I also think that the architecture could be clearer... But these two `addTask` do very different things

`useFirestoreState` provides functions to update the Firestore state.

`useGraphState` provides functions to update the Recoil state

You can see here how the data typically flows:

IMG_1107

File: src/atoms.ts [hunk 3]
```diff
@@ -47,6 +49,11 @@ const dependencyStateFamily = atomFamily<Dependency, DependencyId>({
   },
 });
 
+const projectIdState = atom<ProjectId>({
+  key: "ProjectId",
+  default: "spXxYVulTgfKcj0n1sWb",
```
Comment (justinline) [inline src/atoms.ts:54]: issue: this never changes? unless I missed it somewhere. You're about to share all your personal tasks with the world if you deploy this :D

**Comment by justinline** [comment]
thought: I would personally shelve this until you have all your desired functionality - it's a nice feature but it will create blockers and require more maintenance going forward so it may slow you down. Just my opinion... Is the value of saving on a server so important right now? 
Maybe it's simpler to do a manual save/load, which just copies the JSON state, but you need an authentication flow of some sort first?

**Comment by justinline** [comment]
Finished 🏁 💸

File: src/atoms.ts (hunk 3 above)
Comment (etrobert) [reply src/atoms.ts:54]: Haha in this MR it never changes but I'll add a followup that provides an interface to create a new graph and react routing 😊

File: src/useFirestoreState.ts (hunk 2 above)
Comment (etrobert) [reply src/useFirestoreState.ts:17]: Good point!

**Comment by etrobert** [comment]
> question: what will happen when free limits are hit on firebase?

It happened yesterday, all requests are then denied for the day

File: src/useFirestoreState.ts (hunk 1 above)
Comment (justinline) [reply src/useFirestoreState.ts:15]: hmmm, maybe it might be better to rename the functions like events `userAddedTask` for the UI and keep `addTask` as imperative for firebase?

Comment (etrobert) [reply src/useFirestoreState.ts:15]: That's a good idea, but the problem is that it's not always the case that we go through the above general graph

For example for the position
- We update the Recoil pos directly on every `Draggable.onDrag`
- We only update the Firestore pos on `Draggable.onStop`

So I don't know if it would make much sense to have `Draggable.onDrag` call `userSetTask`

**Comment by etrobert** [comment]
> thought: I would personally shelve this until you have all your desired functionality - it's a nice feature but it will create blockers and require more maintenance going forward so it may slow you down. Just my opinion... Is the value of saving on a server so important right now?
> Maybe it's simpler to do a manual save/load, which just copies the JSON state, but you need an authentication flow of some sort first?

Well, persistence is important

I just removed the save/load from file from master and I am planning to remove the save/load from localStorage asap

Problems with save/load from file & localStorage is that it requires me to be backwards compatible and it is anyway not a behaviour I want eventually in the product...

Although I agree that it is only second in the list of priorities behind "Merging and Releasing the React/Recoil migration" that requires me to migrate the basic functionalities first.

So yes I can wait to merge this MR until this is done 😊

File: src/useFirestoreState.ts (hunk 1 above)
Comment (justinline) [reply src/useFirestoreState.ts:15]: `userUpdatedTask`, no?

File: src/useSyncFirestore.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { useEffect } from "react";
+import { collection, onSnapshot } from "firebase/firestore";
+import { useRecoilValue } from "recoil";
+
+import { projectIdState, Task } from "./atoms";
+import firestore from "./firestore";
+import useGraphState from "./App/useGraphState";
+
+const useSyncFirestore = (): void => {
```
Comment (justinline) [inline src/useSyncFirestore.ts:9]: question: should this logic just live in the `useGraphState` hook? I feel like the syncing can just be done directly there, since this has no other dependencies. Do you see a benefit keeping it separate? I think 3 hooks all dealing with state/firebase is a bit overkill currently and i'm finding it hard to follow :(

File: src/useSyncFirestore.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  const projectId = useRecoilValue(projectIdState);
+  const { addTask, setTask, removeTask } = useGraphState();
+
+  useEffect(() => {
```
Comment (justinline) [inline src/useSyncFirestore.ts:13]: praise: this useEffect is really easy to follow though 👏

File: src/App/useGraphState.ts [hunk 6]
```diff
@@ -1,27 +1,42 @@
...
 type UseGraphState = () => {
-  addTask: (name: string) => void;
+  addTask: (id: TaskId, task: Task) => void;
+  setTask: (id: TaskId, task: Task) => void;
+  removeTask: (id: TaskId) => void;
   clearGraph: () => void;
 };
 
-// TODO Use a real unique id generator
-const generateNewId = () => (Math.random() * 1000).toString();
```
Comment (justinline) [inline src/App/useGraphState.ts:11]: question: we lost the id generation but I assume this is now being handled by firestore?

**Comment by justinline** [comment]
Actually finished now 🏁 really nice integration 👏

File: src/App/useGraphState.ts (hunk 6 above)
Comment (etrobert) [reply src/App/useGraphState.ts:11]: Yes 😊

File: src/useSyncFirestore.ts (hunk 4 above)
Comment (etrobert) [reply src/useSyncFirestore.ts:9]: I'm not saying this is the best architecture in the world.

But `useGraphState` hook returns a set of functions to update the `Recoil` state.

It is currently called in `useSyncFirestore` and in `Task`, and might be called in more places that need to update the Recoil state.

If we included the `useSyncFirestore` inside of `useGraphState` it would add the listeners in all these places, that would not make much sense

If it seems unclear maybe there is a naming/documentation problem

Maybe something like this would be clearer:
```
useFirestoreState - Returns a set of functions to update Firestore
useRecoilState (previously useGraphState) - Returns a set of functions to update Recoil
useSyncFirestore - Sync the Firestore changes in Recoil
```

File: src/useFirestoreState.ts (hunk 1 above)
Comment (etrobert) [reply src/useFirestoreState.ts:15]: `userUpdatedTask` sounds like the name people use in Redux when they have to find another name for the continuation of an action because having two `addTask` actions type would clash

But in this case it's not _always_ one after the other, and even if it was, the `addTask`s are scoped to `useFirestoreState` and `useGraphState` (could become `useRecoilGraphState` or `useRecoilState` or smthing), so there is no name clash problem
